### DISCUSSION TITLE
How does one include additional headers in an F' unit test?

### DISCUSSION THREAD

**Discussion by JackNWhite** [nasa/fprime#1323] [Q&A] [answered]
_Note: Please note that the **full** answer is in the comments for the marked post_

I have inspected the CMake functions, but I confess CMake is completely baffling to me.

I am able to add include paths to my main build with:
```cmake
target_include_directories(${MODULE_NAME}
    ...
    )
```

But I cannot work out the target name for the unit tests. I tried adding `_UT` to the end, but got this error:
```
CMake Error at CMakeLists.txt:48 (target_include_directories):
  Cannot specify include directories for target
  "Something_UT" which is not built by this project.
```
What is the correct procedure?

**Comment by LeStarch** [comment, marked as answer]
@JackNWhite I think you need to try `_ut_exe` e.g. 

```
target_include_dicrectories(${MODULE_NAME}_ut_exe <some directory>)

**Reply by JackNWhite** [reply]
Thanks, I have tried that now, but when running `fprime-util check` from the component it says the project doesn't build that target:
```
Configuring incomplete, errors occurred!
  Cannot specify include directories for target
  "Something_ut_exe" which is not built by this project.
```

**Reply by JackNWhite** [reply]
I've found that I'm able to build if instead of using `target_include_dicrectories` I just used `include_dicrectories`, but I'm unsure of whether that makes my test macros (what I'm trying to include) available to normal sources.

**Reply by LeStarch** [reply]
It would make them available to normal sources.  include_directories affects everything after that point.  Just to clarify, when working with targets created by F´, you should be performing calls after the appropriate register call:

```
get_module_name("${CMAKE_CURRENT_LIST_DIR}")
register_fprime_ut()
target_include_directories("${MODULE_NAME}_ut_exe")
```

**Reply by JackNWhite** [reply]
Thanks for the help with this. Alas, if I add `target_include_directories("${MODULE_NAME}_ut_exe")`, I am unable to generate a normal build, with CMake saying:
```
-- Configuring incomplete, errors occurred!
  Cannot specify include directories for target
See also ".../build-fprime-automatic-native/CMakeFiles/CMakeOutput.log".
  "Something_ut_exe" which is not built by this project.
```

**Reply by LeStarch** [reply]
That one is on me.  I forgot to give you the complete construct.  In F´ and CMake generally, the variable `BUILD_TESTING` is used to turn on and off testing.  In our case, we don't create UT targets unless that is set.  The following will add include directories only in the case that we are testing

```
register_fprime_module()
....
register_fprime_ut()
....
get_module_name("${CMAKE_CURRENT_LIST_DIR}")
if (BUILD_TESTING)
    target_include_directories("${MODULE_NAME}_ut_exe")
endif()
```

**Reply by JackNWhite** [reply]
Sage wisdom! Thank you!
